iceoryx_hoofs/include/iceoryx_hoofs/containers/uninitialized_array.hpp

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_CONTAINERS_UNINITIALIZED_ARRAY_HPP
+#define IOX_HOOFS_CONTAINERS_UNINITIALIZED_ARRAY_HPP
+
+#include "iceoryx_hoofs/iceoryx_hoofs_types.hpp"
+
+#include <cstdint>
+
+namespace iox
+{
+namespace containers
+{
+/// @brief struct used as policy parameter in UninitializedArray to wrap an array with its first element zeroed
+template <typename ElementType, uint64_t Capacity>
+struct FirstElementZeroed
+{
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    // NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+    using element_t = cxx::byte_t[sizeof(ElementType)];
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    alignas(ElementType) element_t value[Capacity]{{0}};
+    // NOLINTEND(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+};
+
+/// @brief struct used as policy parameter in UninitializedArray to wrap an uninitialized array
+template <typename ElementType, uint64_t Capacity>
+struct UninitializedBuffer
+{
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    // NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+    using element_t = cxx::byte_t[sizeof(ElementType)];
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    alignas(ElementType) element_t value[Capacity];
+    // NOLINTEND(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+};
+
+/// @brief Wrapper class for a C-style array of type ElementType and size Capacity. Per default it is uninitialized but
+/// the first element can be zeroed via template parameter FirstElementZeroed.
+/// @note Out of bounds access leads to undefined behavior
+template <typename ElementType, uint64_t Capacity, template <typename, uint64_t> class Buffer = UninitializedBuffer>
+class UninitializedArray
+{
+  public:
+    using iterator = ElementType*;
+    using const_iterator = const ElementType*;
+
+    constexpr UninitializedArray() noexcept = default;
```

Comment:
Oh this is a point I overlooked. The uninitialized array has no knowledge whether a type is constructed and has to be moved/copied or not, this is the responsibility of the user.
So we have to delete copy and move ctor&assignment otherwise we would have to track which element is initialized, which is already done by the constructs using the `uninitialized_array`